src/main/kotlin/org/rust/cargo/project/CargoProjectResolver.kt

```diff
@@ -36,163 +28,149 @@ class CargoProjectResolver : ExternalSystemProjectResolver<CargoExecutionSetting
-
-                moduleOrLibrary.put(pkg.name, moduleNode)
-
-                // Publish source-/test-/resources- roots
-                addContentRoots(moduleNode, root, pkg)
-
-            } else {
-                // Add as a library
-                val libData = LibraryData(CargoProjectSystem.ID, "${pkg.name} ${pkg.version}")
-
-                libData.addPath(LibraryPathType.BINARY, root.absolutePath)
-                libData.addPath(LibraryPathType.SOURCE, root.absolutePath)
-
-                moduleOrLibrary.put(
-                    pkg.name,
-                    projectNode.createChild(ProjectKeys.LIBRARY, libData)
-                )
-            }
+
+        val modules = metadata.modules.map { module ->
+            val moduleNode = createModuleNode(module, projectNode)
+            module.name to moduleNode
         }
 
-        // TODO(winger, kudinkin): add transitive dependencies too?
-
-        // Add dependencies
-        for (pkg in data.packages) {
-            val pkgNode = moduleOrLibrary[pkg.name]!!
-
-            if (pkgNode.key != ProjectKeys.MODULE) {
-                // Skip dependencies for libraries
-                continue
-            }
-
-            for (dep in pkg.dependencies) {
-                val depNode = moduleOrLibrary[dep.name] ?: continue
-
-                if (depNode.key == ProjectKeys.MODULE) {
-                    pkgNode.createChild(
-                        ProjectKeys.MODULE_DEPENDENCY,
-                        ModuleDependencyData(
-                            pkgNode.data as ModuleData,
-                            depNode.data as ModuleData
-                        )
-                    )
-                } else if (depNode.key == ProjectKeys.LIBRARY) {
-                    pkgNode.createChild(
-                        ProjectKeys.LIBRARY_DEPENDENCY,
-                        LibraryDependencyData(
-                                pkgNode.data as ModuleData,
-                                depNode.data as LibraryData,
-                                LibraryLevel.PROJECT
-                        )
-                    )
-                } else {
-                    throw AssertionError("Panic! You may not reach this point!")
-                }
-            }
+        val libraries = metadata.libraries.map { lib ->
+            val libData = LibraryData(CargoProjectSystem.ID, "${lib.name} ${lib.version}")
+            libData.addPath(LibraryPathType.SOURCE, lib.contentRoot.absolutePath)
+            val libNode = projectNode.createChild(ProjectKeys.LIBRARY, libData)
+            lib.name to libNode
         }
 
+        addDependencies(modules, libraries)
+
         return projectNode
     }
 
-    private fun deviseModuleType(pkg: CargoPackageInfo): String {
-        var moduleType: String? = null
+    @Suppress("UNUSED_PARAMETER")
+    private fun addDependencies(modules: List<Pair<String, DataNode<ModuleData>>>,
```

Comment:
I don't see any immediate gain in knowing dependencies between libraries (I may be wrong here), and resolving them properly is a little tricky, especially because we don't have any tests for it. 

By tricky I mean the fact that there can be several versions of a single library.